### PR TITLE
Laikad preperation

### PR DESCRIPTION
**Pull request by gast04**
Contains all changes which are needed to change to laikad.
Does not contain behavioral changes, just prepares for them

### CODE REVIEW

File: selfdrive/locationd/laikad_helpers.py [hunk 1]
```diff
@@ -43,6 +43,67 @@ def Fx_pos(inp, weight=None):
   return Fx_pos
 
 
+def get_prr_sympy_func():
+  # implemting this without sympy.Matrix gives a 2x speedup at generation
```
Comment (sshane) [inline selfdrive/locationd/laikad_helpers.py:47]:
```suggestion
  # implementing this without sympy.Matrix gives a 2x speedup at generation
```

File: selfdrive/locationd/laikad.py [hunk 2]
```diff
@@ -92,20 +91,21 @@ def cache_ephemeris(self, t: GPSTime):
...
+  def get_lsq_fix(self, t, measurements):
+    if self.last_fix_t is None or abs(self.last_fix_t - t) > 0:
+      min_measurements = 6 if any(p.constellation_id == ConstellationId.GLONASS for p in measurements) else 5
+      position_solution, pr_residuals = calc_pos_fix_gauss_newton(measurements, self.posfix_functions, min_measurements=min_measurements)
+      if len(position_solution) < 3:
+        return None
+      position_estimate = position_solution[:3]
+      position_std = np.median(np.abs(pr_residuals)) * np.ones(3)
+      velocity_solution, prr_residuals = calc_vel_fix(measurements, position_estimate, min_measurements=min_measurements)
+      if len(velocity_solution) == 0 or len(prr_residuals) == 0:
```
Comment (HaraldSchafer) [inline selfdrive/locationd/laikad.py:103]: Can you make this check the same as position_solution check? Shouldn't be different?

File: selfdrive/locationd/laikad.py [hunk 3]
```diff
@@ -92,20 +91,21 @@ def cache_ephemeris(self, t: GPSTime):
...
-          cloudlog.debug(f"Pos fix failed with median: {residual_median.round()}. All residuals: {np.round(pos_fix_residual)}")
-    return self.last_pos_fix
+  def get_lsq_fix(self, t, measurements):
+    if self.last_fix_t is None or abs(self.last_fix_t - t) > 0:
+      min_measurements = 6 if any(p.constellation_id == ConstellationId.GLONASS for p in measurements) else 5
+      position_solution, pr_residuals = calc_pos_fix_gauss_newton(measurements, self.posfix_functions, min_measurements=min_measurements)
+      if len(position_solution) < 3:
+        return None
+      position_estimate = position_solution[:3]
+      position_std = np.median(np.abs(pr_residuals)) * np.ones(3)
```
Comment (HaraldSchafer) [inline selfdrive/locationd/laikad.py:101]: Add:
```
#TODO median abs residual is decent estimate of std, can be improved with measurements stds and/or DOP
```

File: selfdrive/locationd/laikad_helpers.py [hunk 4]
```diff
@@ -43,6 +43,67 @@ def Fx_pos(inp, weight=None):
...
+
+  def Fx_vel(vel, no_weight=no_weight):
+    vals, gradients = [], []
+
+    for meas in measurements:
+      if signal not in meas.observables or not np.isfinite(meas.observables[signal]):
+        continue
+
+      sat_pos = meas.sat_pos_final if meas.corrected else meas.sat_pos
+      weight = 1 if no_weight or meas.observables[signal] == 0 else (1 / meas.observables[signal])
```
Comment (HaraldSchafer) [inline selfdrive/locationd/laikad_helpers.py:92]: (1 / meas.observables_std[signal]) !!!

File: selfdrive/locationd/locationd.cc [hunk 5]
```diff
@@ -348,6 +348,77 @@ void Localizer::handle_gps(double current_time, const cereal::GpsLocationData::R
...
+  // indexed at 0 cause all std values are the same MAE
+  auto ecef_pos_std = log.getPositionECEF().getStd()[0];
+  MatrixXdr ecef_pos_R = Vector3d::Constant(pow(ecef_pos_std*10, 2)).asDiagonal();
+
+  auto ecef_vel_v = log.getVelocityECEF().getValue();
+  VectorXd ecef_vel = Vector3d(ecef_vel_v[0], ecef_vel_v[1], ecef_vel_v[2]);
+
+  // indexed at 0 cause all std values are the same MAE
+  auto ecef_vel_std = log.getVelocityECEF().getStd()[0];
+  MatrixXdr ecef_vel_R = Vector3d::Constant(pow(ecef_vel_std*10, 2)).asDiagonal();
```
Comment (HaraldSchafer) [inline selfdrive/locationd/locationd.cc:378]: Let's factor out this ten as a constant up top

Comment (HaraldSchafer) [reply selfdrive/locationd/locationd.cc:378]: It's used in handle_gps too

File: selfdrive/locationd/locationd.cc [hunk 6]
```diff
@@ -348,6 +348,77 @@ void Localizer::handle_gps(double current_time, const cereal::GpsLocationData::R
...
+  for (int i = 0; i < orientation_error.size(); i++) {
+    orientation_error(i) = std::fmod(orientation_error(i), 2.0 * M_PI);
+    if (orientation_error(i) < 0.0) {
+      orientation_error(i) += 2.0 * M_PI;
+    }
+    orientation_error(i) -= M_PI;
+  }
+  VectorXd initial_pose_ecef_quat = quat2vector(euler2quat(ecef_euler_from_ned({ ecef_pos(0), ecef_pos(1), ecef_pos(2) }, orientation_ned_gps)));
+
+  if (ecef_pos_std > 50. || ecef_vel_std > 5.) {
```
Comment (HaraldSchafer) [inline selfdrive/locationd/locationd.cc:401]: Factor out constants

File: selfdrive/locationd/locationd.cc [hunk 7]
```diff
@@ -348,6 +348,77 @@ void Localizer::handle_gps(double current_time, const cereal::GpsLocationData::R
...
+    this->gps_valid = false;
+    this->determine_gps_mode(current_time);
+    return;
+  }
+
+  if ((gps_est_error > 200. && ecef_pos_std < 5.) || this->last_gps_msg == 0) {
```
Comment (HaraldSchafer) [inline selfdrive/locationd/locationd.cc:407]: Magic numbers defined at top

Comment (gast04) [reply selfdrive/locationd/locationd.cc:407]: reset error values are different for GPS and GNSS, i will only define the GNSS values at top, and keep the GPS values in place

File: selfdrive/locationd/laikad.py (hunk 2 above)
Comment (gast04) [reply selfdrive/locationd/laikad.py:103]: changed it, I will see it in the tests if it has influence

File: selfdrive/locationd/laikad_helpers.py (hunk 1 above)
Comment (gast04) [reply selfdrive/locationd/laikad_helpers.py]: ok boss

**Comment by sshane** [comment]
🎉

**Pull request merged**